### ISSUE TITLE
SearchFragment is noticibly slow on first run

### ISSUE THREAD

**Issue by Baron-Severin** [mozilla-mobile/fenix#6274]
## Steps to reproduce
- Open a fresh Fenix instance
- Click on the URL bar

### Expected behavior
- Search fragment should appear quickly

### Actual behavior
- Search fragment is noticeably slow to appear
- This only happens the first time the SearchFragment is launched

### Device information

* Android device: physical Pixel 2, emulator
* Fenix version: 9d91108f4e6576860bc1da734768b4422d08cba8 (current head of master)

### Notes
I didn't spend much time on this, but on my emulator the first launch (from URL bar click to onViewCreated) took about 160 MS.  Follow up launches varied, but hovered around 30-60 MS.  I believe the animated transition makes these numbers look worse than they are (follow up launches feel quite snappy).

**Comment by emalysz** [comment]
After capturing this in the profiler, SearchFragment's onCreateView() indeed takes longer after first run. First run: 1,048,032 μs. Second run: 322,882 μs.
Breaking this down, it looks like the increase is mostly due to the time spend in <init> for both ToolbarView and AwesomeBarView.

**Comment by hakkikaancaliskan** [comment]
This problem looks like fixed for me after
Nightly 191128 11:47 (Build #13321153)
📦: 24.0.0, 3183401f0
🦎: 72.0a1-20191126093448
XZ1, Android 10
Can someone confirm? Or initializing megazord asynchronously meet the required time of searchfragment?

**Comment by Baron-Severin** [comment]
@boek and I were still able to reproduce this, although it's less noticeable on newer devices.  We did some investigation and it looks like there's also a very substantial amount of time spent in the `SearchFragment` `consumeBy(searchStore)` block.  This might be easier for us to fix since view init is handled in AC.  Once we have a better idea of what's going on, we can open an AC issue as well.

**Comment by Baron-Severin** [comment]
It may be possible to fix the perceived perf issue here without modifying toolbar internals by [inflating the toolbar on a background thread](https://developer.android.com/reference/android/support/v4/view/AsyncLayoutInflater.html) on app init.

**Comment by hakkikaancaliskan** [comment]
> 
> 
> It may be possible to fix the perceived perf issue here without modifying toolbar internals by [inflating the toolbar on a background thread](https://developer.android.com/reference/android/support/v4/view/AsyncLayoutInflater.html) on app init.

[Androidx](https://developer.android.com/reference/kotlin/androidx/asynclayoutinflater/view/AsyncLayoutInflater.html?hl=en) page of async layout inflater. Just wanted to help ¯\_(ツ)_/¯.

**Comment by MarcLeclair** [comment]
( I just saw this issue), and I came across a similar problem where the initialization of the ToolBar took ~500ms-1 second (it was varying a lot) when creating a new tab and navigating to a website.  I was wondering if using AsyncLayoutInflate could be used there too?  (I'll add my trace for reference -- check from timestamp 27 second until 32 second, thats when I entered the URL search bar until the webpage fully loaded) 
[new_tab.trace.zip](https://github.com/mozilla-mobile/fenix/files/3975622/new_tab.trace.zip)

**Comment by Baron-Severin** [comment]
Wow, good catch @MarcLeclair.  It looks like you're already involved in discussion on this in another issue, linking here for future readers.

[[Bug] Long delay when loading new tab](https://github.com/mozilla-mobile/fenix/issues/7143)

**Comment by colintheshots** [comment]
This is likely fixed by the same PR @hawkinsw posted for #7143. Will can close this also if it's fixed.

**Comment by hakkikaancaliskan** [comment]
> This is likely fixed by the same PR @hawkinsw posted for #7143. Will can close this also if it's fixed.

What about @Baron-Severin's offer about AsyncLayoutInflater?
There will be a good perf gain.

**Comment by Baron-Severin** [comment]
I'd like to remeasure with @hawkinsw's fix, first.  Async inflation is definitely a viable approach here, but it will add some complexity to code that's already fairly difficult to work with.  If we see substantially better performance with this much simpler solution, I'd be happy to stick with that.

**Comment by MarcLeclair** [comment]
Also, @Baron-Severin  I noticed that using AsyncLayoutInflater may cause trouble in some part of our code since we rely heavily on inflation of one part to create another. For instance, I tried it out with our DefaultBrowserToolbar and the inflation is somewhat sequential to inflate another (especially since we use a ViewStub then create one after the other). I'm not too sure how that would work out in some part of our code. Just  food for thought I guess 😄

**Comment by hawkinsw** [comment]
At first I didn't see how #7143 is related to this. However, I think that I see it now:
1. Click the Awesome Bar
2. See the Awesome Bar UI appear (no delay here)
3. Enter search term(s)
4. Click Enter
5. See the Search Fragment Appear (delay here) 

Please confirm that this is the use case. 

*If* that is the correct user scenario then (based on the great trace from @MarcLeclair ) these are identical issues. 

Would you be willing to confirm, @Baron-Severin? Thanks :-)

**Comment by hawkinsw** [comment]
Added @Baron-Severin as an assignee because github doesn't have an needinfo flag! 

Bugzilla FTW!

**Comment by Baron-Severin** [comment]
@hawkinsw nope, this issue is about a delay on step 2.  I checked out your branch and it's still reproducible there, so it looks like you're right that it's unrelated.

AFAICT, the only options here are 1) async inflation, or 2) a deep dive into the AC toolbar implementation.  Unless you see a better option @colintheshots.

**Comment by MarcLeclair** [comment]
So , I gave it a little test run (I've never really used asyncLayoutInflater so my implementation might be somewhat off -- especially since I'm not sure what everything does). But, the "preliminary" results seem promising (refer to here: [issue 6274 in my fork](https://github.com/MarcLeclair/fenix/tree/issue_6274_asynclayoutinflater_for_toolbar))

TLDR:  Without async layout inflate the searchFragment took ~574 ms 
             With the async layout inflate the searchFragment took ~276 ms. 

It is much faster but some of the UI behaves weirdly (like the hint for the url bar, and I didn't try when on a web page, I just did it to see early speed).

Traces for the interested:
[with and without async traces.zip](https://github.com/mozilla-mobile/fenix/files/3997283/with.and.without.async.traces.zip)

Timeline of each trace:
         1.  _with_asyncInflater.trace_ : 
              - Touch event happens at ~8.6 seconds 
              - fragment appears at ~9.5 seconds ( you can see the asyncInflate takes ~3.56ms for the toolbar)
         2. _without_async.trace_ : 
              - Touch event happens at ~10.4 seconds 
              -  fragment appears 11.5 seconds

**Comment by MarcLeclair** [comment]
To answer my earlier comment for future reference: the Async loading seemed to affect the bitmap loading as well, so bitmaps should tried to be precomputed or something before calling our async (just did a push that did something like that, i guess?)

**Comment by Baron-Severin** [comment]
That's good progress!  I haven't had time to really dive into this yet, but I was hoping we might be able to eagerly inflate the `SearchFragment` on a background thread during application init.  Then on search bar click we block until the fragment is ready.  Hopefully we would be able to get rid of the delay completely for the large majority of cases.  I haven't looked into how feasible this is yet, though.

**Comment by hawkinsw** [comment]
Hey @MarcLeclair - I tested this and also see a performance improvement. I left a few comments on your branch. They are mostly things that I am confused/intrigued about. Thanks for doing this great work!

**Comment by hawkinsw** [comment]
Putting @MarcLeclair as an assignee because, again, github does not have a needinfo flag.

**Comment by hawkinsw** [comment]
I think that I profiled this previously and I think it is related to Gecko session initialization. I will take another look later tonight and start to do more thorough investigation.

**Comment by hawkinsw** [comment]
I spent a little time tonight working on an implementation that uses an asynclayoutinflater triggered by the start of the Home Fragment and it seems to work. The delay is significantly improved and there is still more that we could do. I will work with @MarcLeclair and @Baron-Severin on Monday to see if the approach I took is something that we think could work.

**Comment by mcomella** [comment]
After Will removed the `WebURLFinder` regex in this startup path, the STR in this issue is not noticeable on my Pixel 2 Nightly build and is barely noticeable on my performanceTest (i.e. debuggable) build. Similarly, it's barely noticeable on my G5 non-debuggable build.

I think we've done enough to qualify for "no longer a P1 responsiveness issue" but should file a follow-up to address the sluggishness that remains post-GA.

**Comment by mcomella** [comment]
@Baron-Severin How do you feel about the changes? Can you still reproduce the issue?

If not and @hawkinsw agrees there is nothing else to be done here, I suggest we close after filing the follow-up.

**Comment by Baron-Severin** [comment]
@mcomella I can still reproduce on nightly, but it's _greatly_ mitigated. Definitely no longer a P1. I think we should land the component because imo it's a valid approach that we should consider using in some places, but it's probably no longer worth the complexity here.

**Comment by mcomella** [comment]
Let's close and we can revisit whenever we prioritize the search experience for perf evaluation (like we're doing for cold start right now).

@hawkinsw Can you add a QA note?
